drivers/meter/ade7880/ade7880.c

```diff
@@ -0,0 +1,487 @@
+		buff[3] = (uint8_t)reg_data;
+
+		return no_os_spi_write_and_read(dev->spi_desc, buff, 4);
+	}
+
+	/* 16 bits registers */
+	if ((reg_addr >= ADE7880_REG_PHSTATUS && reg_addr <= ADE7880_REG_CONFIG) ||
+	    (reg_addr >= ADE7880_REG_HCONFIG && reg_addr <= ADE7880_REG_LAST_RWDATA16) ||
+	    (reg_addr == ADE7880_REG_RUN)) {
+		no_os_put_unaligned_be16(reg_data, &buff[3]);
+
+		return no_os_spi_write_and_read(dev->spi_desc, buff, 5);
+	}
+
+	/* 32 bits registers */
+	no_os_put_unaligned_be32(reg_data, &buff[3]);
+
+	return no_os_spi_write_and_read(dev->spi_desc, buff, 7);
+}
+
+/**
+ * @brief Set power mode
+ * @param dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade7880_set_power_mode(struct ade7880_dev *dev)
+{
+	int ret;
+
+	if (!dev)
+		return -ENODEV;
+	if (!dev->psm0_desc)
+		return -EINVAL;
+	if (!dev->psm1_desc)
+		return -EINVAL;
+
+	switch (dev->power_mode) {
+	case SLEEP_MODE:
+		ret = no_os_gpio_set_value(dev->psm1_desc,
+					   NO_OS_GPIO_HIGH);
+		if (ret)
+			return ret;
+		ret = no_os_gpio_set_value(dev->psm0_desc,
+					   NO_OS_GPIO_HIGH);
+		break;
+	case LOW_POWER_MODE:
+		ret = no_os_gpio_set_value(dev->psm1_desc,
+					   NO_OS_GPIO_HIGH);
+		if (ret)
+			return ret;
+		ret = no_os_gpio_set_value(dev->psm0_desc,
+					   NO_OS_GPIO_LOW);
+		break;
+	case REDUCED_POWER_MODE:
+		ret = no_os_gpio_set_value(dev->psm1_desc,
+					   NO_OS_GPIO_LOW);
+		if (ret)
+			return ret;
+		ret = no_os_gpio_set_value(dev->psm0_desc,
+					   NO_OS_GPIO_LOW);
+		break;
+	case NORMAL_MODE:
+		ret = no_os_gpio_set_value(dev->psm1_desc,
+					   NO_OS_GPIO_LOW);
+		if (ret)
+			return ret;
+		ret = no_os_gpio_set_value(dev->psm0_desc,
+					   NO_OS_GPIO_HIGH);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Update specific register bits.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param mask - Specific bits mask.
+ * @param reg_data - The data to be written.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade7880_update_bits(struct ade7880_dev *dev, uint16_t reg_addr,
+			uint32_t mask, uint32_t reg_data)
+{
+	int ret;
+	/* register value read */
+	uint32_t data;
+
+	if (!dev)
+		return -ENODEV;
+
+	ret = ade7880_read(dev, reg_addr, &data);
+	if (ret)
+		return ret;
+
+	data &= ~mask;
+	data |= reg_data & mask;
+
+	return ade7880_write(dev, reg_addr, data);
+}
+
+/**
+ * @brief Get interrupt indicator from STATUS0 register.
+ * @param dev - The device structure.
+ * @param msk - Interrupt mask.
+ * @param status - Status indicator.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade7880_get_int_status0(struct ade7880_dev *dev, uint32_t msk,
+			    uint8_t *status)
+{
+	int ret;
+	/* register value read */
+	uint32_t reg_val;
+
+	if (!dev)
+		return -ENODEV;
+	if (!status)
+		return -EINVAL;
```

Comment:
You could skip these checks, as they are already done in the `ade7880_read()` function.